src/eterna/folding/EternafoldThreshknot.ts

```diff
@@ -0,0 +1,172 @@
+            const dotPlotResult = this._lib.GetDotPlot(_temp, seqStr);
+            bpp = EmscriptenUtil.stdVectorToArray(dotPlotResult.plot);
+
+            // THRESHKNOT HEURISTIC
+            // Initialize algorithm parameters, iteration flags, and base pair list variable
+            const maxIterations = 3;
+            // TODO: Implement in SecStruct.stems()
+            // Currently stems() interprets bulges as stem breaks
+            // const allowedBulgeLen = 0;
+            const minLenHelix = 2;
+            let iteration = 0;
+            let newBP = 1;
+            const bpList: number[][] = [];
+
+            // Get pairs for each base via Threshknot heuristic
+            while (newBP !== 0 && iteration <= maxIterations) {
+                const current_bpList: number[][] = [];
+                const bpListFlat = bpList.flat();
+                const Pmax = new Array(seqStr.length).fill(0);
+
+                // Find max probability for each base, filtering out already paired bases
+                for (let index = 0; index < bpp.length; index += 3) {
+                    const base1 = bpp[index];
+                    const base2 = bpp[index + 1];
+                    const prob = bpp[index + 2];
+                    // console.log(base1, base2, prob, !(bpListFlat.includes(base1) || bpListFlat.includes(base2)));
+                    if (!(bpListFlat.includes(base1) || bpListFlat.includes(base2))) {
+                        // -1 to account for 0 indexing
+                        if (prob >= Pmax[base1 - 1]) Pmax[base1 - 1] = prob;
+                        if (prob >= Pmax[base2 - 1]) Pmax[base2 - 1] = prob;
+                    }
+                }
+
+                // Compare each COO tuple to Pmax and theta, and add to the list
+                // of selected base pairs if the check passes
+                for (let index = 0; index < bpp.length; index += 3) {
+                    const base1 = bpp[index];
+                    const base2 = bpp[index + 1];
+                    const prob = bpp[index + 2];
+                    if (prob === Pmax[base1 - 1] && prob === Pmax[base2 - 1] && prob > _theta) {
+                        current_bpList.push([base1, base2]);
+                    }
+                }
+                // Update iteration flags
+                newBP = current_bpList.length;
+                iteration += 1;
+                if (newBP !== 0 && iteration > maxIterations) {
+                    log.debug('Reached max iteration, stopping before converged.');
+                } else {
+                // Add selected base pairs to output list
+                    bpList.push(...current_bpList);
+                }
+            }
+
+            // Ensure that the bpList is sorted, then check for duplicated nucleotides
+            bpList.sort((bpA, bpB) => bpA[0] - bpB[0]);
+            const nts = bpList.flat();
+            if (nts.length > Array.from(new Set(nts)).length) {
```

Comment:
It should be possible to use Set#size instead of converting it back to an array to get its length

Reply:
Good catch